src/frontend/Desugar.ml

```diff
@@ -0,0 +1,97 @@
+open Core_kernel
+open Ast
+
+let is_multi_index = function
+  | Single {Ast.emeta= {Ast.type_= UArray _; _}; _}
+   |Downfrom _ | Upfrom _ | Between _ | All ->
+      true
+  | Single _ -> false
+
+let remove_trailing_alls_expr = function
+  | Indexed (obj, indices) ->
+      let rec remove_trailing_alls indices =
+        match List.rev indices with
+        | All :: tl -> remove_trailing_alls (List.rev tl)
+        | _ -> indices
+      in
+      Indexed (obj, remove_trailing_alls indices)
+  | e -> e
+
+(* TODO: same thing for Assignments *)
```

Comment:
It's very easy to also do this for assignments: there are functions for converting an lvalue back and forth to an expression.